README.md
CommonPy<img width="12%" align="right" src="https://github.com/caltechlibrary/commonpy/raw/main/.graphics/commonpy-icon.png">
===============================================

This is a collection of common utility functions and classes that we at the Caltech Library have found useful in our other Python projects.

[![Latest release](https://img.shields.io/github/v/release/caltechlibrary/commonpy.svg?style=flat-square&color=b44e88&label=Latest%20release)](https://github.com/caltechlibrary/commonpy/releases)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg?style=flat-square)](https://choosealicense.com/licenses/bsd-3-clause)
[![Python](https://img.shields.io/badge/Python-3.6+-brightgreen.svg?style=flat-square)](http://shields.io)
[![PyPI](https://img.shields.io/pypi/v/commonpy.svg?style=flat-square&color=orange&label=PyPI)](https://pypi.org/project/commonpy/)

Table of contents
-----------------

* [Introduction](#introduction)
* [Installation](#installation)
* [Usage](#usage)
* [Getting help](#getting-help)
* [Contributing](#contributing)
* [License](#license)
* [Authors and history](#authors-and-history)
* [Acknowledgments](#authors-and-acknowledgments)


Introduction
------------

This repository does not constitute a single program; instead, it contains a collection of modules with utility functions and classes that we have found ourselves using repeatedly in other Python projects.


Installation
------------

The instructions below assume you have a Python interpreter installed on your computer; if that's not the case, please first [install Python version 3](INSTALL-Python3.md) and familiarize yourself with running Python programs on your system.

On **Linux**, **macOS**, and **Windows** operating systems, you should be able to install `commonpy` with [`pip`](https://pip.pypa.io/en/stable/installing/).  To install `commonpy` from the [Python package repository (PyPI)](https://pypi.org), run the following command:
```
python3 -m pip install commonpy
```

As an alternative to getting it from [PyPI](https://pypi.org), you can use `pip` to install `commonpy` directly from GitHub, like this:
```sh
python3 -m pip install git+https://github.com/caltechlibrary/commonpy.git
```


Usage
-----

The basic approach to using this package is to import the modules and functions you need.  For example:

```python
from commonpy.file_utils import readable

if readable('/path/to/some/file'):
    # do something
```

The following subsections describe the different modules available.


### Data structures

The `data_structures` module provides miscellaneous data classes.

| Class          | Purpose |
|----------------|---------|
| `CaseFoldDict` | A version of `dict` that compares keys in a case-insensitive manner |
| `CaseFoldSet`  | A version of `set` that compares keys in a case-insensitive manner |


### Data utilities

The `data_utils` module provides a number of miscellaneous simple functions for some common operations on data of various kinds.

| Function           | Purpose |
|--------------------|---------|
| `expanded_range(string)` | Given a string of the form "X-Y", returns the list of integers it represents |
| `flattened(thing)` | Takes a list or dictionary and returns a recursively flattened version |
| `ordinal(integer)` | Returns a string with the number followed by "st", "nd, "rd", or "th" |
| `parsed_datetime(string)` | Returns a date object representing the given date string |
| `pluralized(word, n, include_num)`  | Returns a plural version of `word` if `n > 1` |
| `sliced(list, n)`  | Yields `n` number of slices from the `list` |
| `timestamp()`      | Returns a string for an easily-readable form of the current time and date |
| `unique(list)`     | Takes a list and return a version without duplicates |


### File utilities

The `file_utils` module provides a number of miscellaneous simple functions for some common operations on files and directories.

| Function           | Purpose |
|--------------------|---------|
| `alt_extension(file, ext)` | Returns `file` with the extension replaced by `ext` |
| `copy_file(src, dst)` | Copies file from `src` to `dst` |
| `delete_existing(file)` | Deletes the given `file` |
| `filename_basename(file)` | Returns `file` without any extensions |
| `filename_extension(file)` | Returns the extension of filename `file` |
| `files_in_directory(dir, ext, recursive)` | |
| `filtered_by_extensions(list, endings)` | |
| `nonempty(file)`   | Returns `True` if file `file` is not empty |
| `open_file(file)` | Opens the `file` by calling the equivalent of "open" on this system |
| `open_url(url)` | Opens the `url` in the user's default web browser |
| `readable(dest)`   | Returns `True` if file or directory `dest` is accessible and readable |
| `relative(file)`   | Returns a path string for `file` relative to the current directory |
| `rename_existing(file)` | Renames `file` to `file.bak` |
| `writable(dest)`   | Returns `True` if file or directory `dest` can be written |


### Interruptible wait and interruption handling utilities

The `interrupt` module includes `wait(...)`, a replacement for `sleep(...)` that is interruptible and works with multiple threads.  It also provides methods to cause an interruption (including doing it by issuing a <kbd>^C</kbd> to the program), check whether an interruption occurred, and other related operations.

| Function                 | Purpose |
|--------------------------|---------|
| `config_interrupt(callback, raise_ex, signal)` | Sets up a callback function |
| `interrupt()`            | Interrupts any `wait` in progress |
| `interrupted() `         | Returns `True` if an interruption has been called |
| `raise_for_interrupts()` | Raises an exception if `interrupt()` has been invoked |
| `reset_interrupts()`     | Resets the interruption flag |
| `wait(duration)`         | Waits for `duration` in an interruptible fashion |


### Module utilities

The `module_utils` collection of functions is useful for working with paths related to a running module, for example to find internal data files that might be needed for normal operation.

| Function           | Purpose |
|--------------------|---------|
| `config_path(module_name)` | Returns the path to local config data directory for the module |
| `datadir_path(module_name)` | Returns the path to the `/data` subdirectory of the module |
| `desktop_path()`   | Returns the path to the user's Desktop directory on this system |
| `installation_path(module_name)` | Returns the path to module's installation directory |
| `module_path(module_name)` | Returns the path to the installed module |

Function `config_path(...)` is useful to use in conjunction with Python's [`configparser`](https://docs.python.org/3/library/configparser.html) module.  It returns `~/.config/modulename/` on Unix-like systems.


### Network utilities

The `network_utils` module provides several functions that are useful when performing network operations.

| Function                         | Purpose                                                             |
| -------------------------------- | ------------------------------------------------------------------- |
| `download(url, local_dest)`      | Download a file                                                     |
| `download_file(url, local_dest)` | Download a file without raising exceptions                          |
| `hostname(url)`                  | Returns the hostname portion of a URL                               |
| `net(...)`                       | See below                                                           |
| `netlock(url)`                   | Returns the hostname, port number (if any), and login info (if any) |
| `network_available()`            | Returns `True` if external hosts are reacheable over the network    |
| `on_localhost(url)`              | Returns `True` if the address of `url` points to the local host     |
| `scheme(url)`                    | Returns the protocol portion of the url; e.g., "https"              |


#### _`net`_

The `net` function in the `network_utils` module implements a fairly high-level network operation interface that internally handles timeouts, rate limits, polling, HTTP/2, and more. The function signature is:

```python
net(method, url, client = None, handle_rate = True, polling = False, recursing = 0, **kwargs)
```

The `method` parameter should have a value such as `'get'`, `'post'`, `'head'`, or similar.  The function returns a tuple of (response, exception), where the first element is the response from the get or post HTTP call, and the second element is an exception object if an exception occurred.  If no exception occurred, the second element will be `None`.  This allows the caller to inspect the response even in cases where exceptions are raised.

If keyword `client` is not `None`, it's assumed to be a [Python HTTPX Client](https://www.python-httpx.org) object to use for the network call.  Settings such as timeouts should be done by the caller creating appropriately-configured [Client](https://www.python-httpx.org/api/#client) objects.

If keyword `handle_rate` is `True`, this function will automatically pause and retry if it receives an HTTP code 429 ("too many requests") from the server.  If `False`, it will return the exception `CommonPy.RateLimitExceeded` instead.

If keyword `polling` is `True`, certain statuses like 404 are ignored and the response is returned; otherwise, they are considered errors.  The behavior when `True` is useful in situations where a URL does not exist until something is ready at the server, and the caller is repeatedly checking the URL.  It is up to the caller to implement the polling schedule and call this function (with `polling = True`) as needed.

This method always passes the argument `allow_redirects = True` to the underlying Python HTTPX library network calls.


#### _`download` and `download_file`_

The functions `download(url, local_destination)` and `download_file(url, local_destination)` download a file at the given `url`, writing it to the file specified by the parameter `local_destination`. The former version of the function will raise exceptions in case of problems; the latter version simply return `True` or `False` depending on the success of the download.


### String utilities

| Function           | Purpose |
|--------------------|---------|
| `antiformat(s)`    | Quote instances of `{` and `}` in `s` so it can be passed to format. |
| `print_boxed(msg)` | Print a message with a box around it using pure ASCII characters. |


### System utilities

| Function           | Purpose |
|--------------------|---------|
| `system_profile()` | Returns a string describing the system running this Python program. |


### Exceptions

The CommonPy module defines a number of exceptions that it may return.  (Most of the exceptions are potentially thrown by `net`, discussed above.)

| Exception                | Meaning |
|--------------------------|---------|
| `CommonPyException`      | Base class for CommonPy exceptions |
| | |
| `ArgumentError`          | The function call was given invalid or unexpected arguments |
| `AuthenticationFailure`  | Problem obtaining or using authentication credentials |
| `InternalError`          | Unrecoverable problem involving CommonPy itself |
| `Interrupted`            | The user elected to cancel/quit the program |
| `NetworkFailure`         | Unrecoverable problem involving net | 
| `NoContent`              | No content found at the given location |
| `RateLimitExceeded`      | The service flagged reports that its rate limits have been exceeded |
| `ServiceFailure`         | Unrecoverable problem involving a remote service |


Getting help
------------

If you find an issue, please submit it in [the GitHub issue tracker](https://github.com/caltechlibrary/commonpy/issues) for this repository.


Contributing
------------

We would be happy to receive your help and participation with enhancing CommonPy!  Please visit the [guidelines for contributing](CONTRIBUTING.md) for some tips on getting started.

If you plan on doing any development on CommonPy, you may want to install the package dependencies listed in [`requirements-dev.txt`](requirements-dev.txt), e.g., using a command such as the following. This will install dependencies necessary to run `pytest`.
```
python3 -m pip install -r requirements-dev.txt
```


License
-------

Software produced by the Caltech Library is Copyright (C) 2020-2023, Caltech.  This software is freely distributed under a BSD/MIT type license.  Please see the [LICENSE](LICENSE) file for more information.


Authors and history
---------------------------

Mike Hucka started this collection of utilities sometime in 2018.


Acknowledgments
---------------

This work was funded by the California Institute of Technology Library.


CHANGES.md
# Change log for CommonPy

## Version 1.13.0

Changes in this release:
* HTTP code 422 is now more accurately returned as a `ServiceFailure` error instead of a `NetworkFailure`.


## Version 1.12.4

Changes in this release:
* Revise the retry algorithm in `timed_request()` again.
* Make `timed_request()` print the response text in case of failures
* Remove `extra_requires` from `setup.py` because it caused installation problems due to the syntax of the requirements file and I don't have patience for figuring out what `setuptools` needs this time.


## Version 1.12.3

This release updates the version of the `dateparser` package in `requirements.txt`, to avoid a deprecation warning when using CommonPy's `parsed_datetime(...)` function.


## Version 1.12.2

This release updates the versions of dependencies in `requirements.txt`.


## Version 1.12.1

This version prevents certain exceptions from being buried and ignored, and improves the network failure retry algorithm.


## Version 1.12.0

Additions in this release:
* New function `network` in the `network_utils` module. It is a companion to `net` and takes the same arguments, but returns only one value (the response). If an error occurs, it raises the error as an exception. This makes it possible for callers to use `network(...)` in somewhat more Pythonic style than `net(...)`,  by wrapping the call to `network(...)` in `try`-`except`.

Changes in this release:
* Removed `slice` from `data_utils` module because it shadows a Python built-in.
* Fixed `hostname`  in `network_utils` to be more general and not hardwire a test for `http`.
* Fixed a bunch of `flake8` warnings.


## Version 1.11.0

Additions in this release:
* New class `CaseFoldSet`, similar to `CaseFoldDict` but … a set.

Changes in this release:
* Fixed a bug in the class documentation in the `README.md` file.
* Added missing dependency for [twine]() in requirements-dev
* Now using lazy `import`s in more places, for faster load times.


## Version 1.10.0

Changes in this release:

* `data_utils.flattened` now outputs `[]` as the value of dict or mapping keys whose original values are an empty sequence (e.g., when the value of a dict key is `[]`). Previously, it would output `None` as the value, which was an unexpected transformation of the input.

Bug fixes in this release:

* Fixed a bug in `data_utils.flattened` that caused it to ignore the `separator` argument in some cases.
* Fixed a missing import of `freezegun` in `requirements-dev.txt`.
* Pin the imported version of regex 2022.3.2, because more recent versions cause calls to `dateparser` to encounter an error ("regex._regex_core.error: bad escape \d at position 7").


## Version 1.9.5

The main functional change in this release is that error objects returned by `net(...)` will have error message bodies returned by network services, where possible.

Internally, there has been some refactoring of the Makefile and addition of things like GitHub template files.


## Version 1.9.4

This release merely changes the version of httpx required by `requirements.txt`. No other changes.


## Version 1.9.3

This version adds a missing internal function definition in `download(...)`.


## Version 1.9.2

This version changes `pluralized` so that it outputs `0 items` instead of `0 item` if the number of its argument has length 0.


## Version 1.9.1

This version adds more test cases and splits out the requirements for testing/development into a separate requirements file, [`requirements-dev.txt`](requirements-dev.txt).


## Version 1.9.0

Changes in this release:
* New exception `ArgumentError`.
* `net(...)` makes slightly finer distinctions between `ServiceFailure` and `NetworkFailure` when it comes to addresses it can't connect to, and also raises `ArgumentError` in some cases such as passing a `data` keyword argument to a network `get`.
* `on_localhost(...)` logs slightly better debug messages.
* There are more test cases.
* `CITATION.cff` has been corrected and no longer describes the wrong repository.


## Version 1.8.2

This version merely changes a version dependency in `requirements.txt`. No other changes.


## Version 1.8.1

Changes in this release:

* At some point between versions 0.18 and 0.21.1 (current version), the `httpx` package changed the keyword argument named `allow_redirects` to `follow_redirects`, and our `network_utils` module functions broke as a result. This release updates `network_utils` to account for the change.
* The `requirements.txt` file now pins most dependencies to a specific version, to avoid situations where getting a newer version of a package might break existing code.
* The copyright year has been updated in various files.


## Version 1.8.0

This version introduces the `data_structures` module, and a new class, `CaseFoldDict`.


## Version 1.7.3

Changes in this release:

* Fixed a missing f-string in an exception message, plus guarded a few more exception message strings with `antiformat`.
* Added `CITATION.cff` file.
* Updated the `Makefile`.


## Version 1.7.2

Improve `flattened` to handle lists of dict keys.


## Version 1.7.1

Add missing `deprecation` package to `requirements.txt`.


## Version 1.7.0

Changes in this release:

* Beginning with this release, the function `reset()` is deprecated in favor of the function `reset_interrupts()`. 
* The function `wait()` resets the interrupt state before it begins waiting.


## Version 1.6.3

This release fixes some unexpected failures in `flattened(...)`.


## Version 1.6.2

This release expands `flattened(...)` to deal with iterators and generators.


## Version 1.6.1

This release fixes a bug in version 1.6.0.


## Version 1.6.0

This release adds the new function `flattened(...)`


## Version 1.5.0

This release adds the new function `print_boxed(...)`.


## Version 1.4.0

Changes in this version:

commonpy/network_utils.py
'''
network_utils.py: network utilities

Authors
-------

Michael Hucka <mhucka@caltech.edu> -- Caltech Library

Copyright
---------

Copyright (c) 2018-2023 by the California Institute of Technology.  This code
is open-source software released under a 3-clause BSD license.  Please see the
file "LICENSE" for more information.
'''

from   ipaddress import ip_address
from   os import stat
import socket
import urllib.parse

if __debug__:
    from sidetrack import log

from .interrupt import wait, interrupted, raise_for_interrupts
from .exceptions import ArgumentError, Interrupted, InternalError, NoContent
from .exceptions import AuthenticationFailure, ServiceFailure, NetworkFailure
from .exceptions import RateLimitExceeded
from .string_utils import antiformat


# Internal constants.
# .............................................................................

_MAX_RECURSIVE_CALLS = 10
'''How many times can certain network functions call themselves upon
encountering a network error before they stop and give up.'''

_MAX_CONSECUTIVE_FAILS = 3
'''Maximum number of times we back off and try again.  This also affects the
maximum wait time that will be reached after repeated retries.'''

_KNOWN_HTTP_METHODS = ['get', 'post', 'head', 'options', 'put', 'delete', 'patch']
'''Known http methods.'''


# Main functions.
# .............................................................................

def network_available(address="8.8.4.4", port=53, timeout=5):
    '''Return True if it appears we have a network connection, False if not.
    By default, this attempts to contact one of the Google DNS servers (as a
    plain TCP connection, not as an actual DNS lookup).  Argument 'address'
    and 'port' can be used to test a different server address and port.  The
    socket connection is attempted for 'timeout' seconds.
    '''
    # Portions of this code are based on the answer by user "7h3rAm" posted to
    # Stack Overflow here: https://stackoverflow.com/a/33117579/743730
    try:
        if __debug__: log('testing if we have a network connection')
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((address, port))
        if __debug__: log('we have a network connection')
        return True
    except (socket.error, socket.timeout):
        if __debug__: log(f'could not connect to {address} in {timeout} sec')
        return False


def hostname(url):
    parts = urllib.parse.urlsplit(url)
    if parts.scheme:
        return parts.hostname
    else:
        # urllib.parse doesn't provide a hostname.  Try a different way.
        import tldextract
        return '.'.join(part for part in tldextract.extract(url) if part)


def scheme(url):
    parsed = urllib.parse.urlsplit(url)
    return parsed.scheme


def netloc(url):
    parsed = urllib.parse.urlsplit(url)
    if parsed.netloc:
        return parsed.netloc
    elif not url.startswith('http') and '//' not in url:
        # Add a fake "http://" so that urllib.parse can figure out the netloc.
        parsed = urllib.parse.urlsplit('http://' + url)
        return parsed.netloc
    else:
        # Last-ditch effort.
        return parsed.path


def on_localhost(url):
    if not url:
        return False
    host = hostname(url)
    # The following approach is based on code posted by user "georgexsh" on
    # 2017-12-21 to https://stackoverflow.com/a/47919356/743730
    for family in (socket.AF_INET, socket.AF_INET6):
        try:
            addrinfo = socket.getaddrinfo(host, None, family, socket.SOCK_STREAM)
        except socket.gaierror:
            break
        else:
            for _, _, _, _, sockaddr in addrinfo:
                address_part = sockaddr[0]
                if ip_address(address_part).is_loopback:
                    if __debug__: log(f'address seems to be on localhost: {url}')
                    return True
    if __debug__: log(f'address not on localhost: {url}')
    return False


def timed_request(method, url, client=None, **kwargs):
    '''Perform a network access, automatically retrying if exceptions occur.

    The value given to parameter "method" must be a string chosen from among
    valid HTTP methods, such as "get", "post", or "head".  If "client" is
    not None, it is used as an httpx.Client object. Other keyword arguments
    are passed to the network method.

    If not given a Client object, the default timeouts for network connect,
    read, and write are 15 seconds.  It enables HTTP 2.0 and disables SSL
    verification.

    This method retries connections in cases of network exceptions.  It also
    retries connections one time when the server returns certain HTTP status
    codes, specifically 400, 409, 502, 503, and 504.  These are sometimes the
    result of temporary server problems or other issues and disappear when a
    second attempt is made after a brief pause.
    '''
    import httpx

    def addurl(text):
        return f'{text} for {url}'

    if client is None:
        timeout = httpx.Timeout(15, connect=15, read=15, write=15)
        client = httpx.Client(timeout=timeout, http2=True, verify=False)
    elif client == 'stream':
        client = httpx.stream

    response = None
    failures = 0
    error = None
    while failures <= _MAX_CONSECUTIVE_FAILS and not interrupted():
        try:
            if __debug__: log(addurl(f'doing http {method}'))
            func = getattr(client, method)
            response = func(url, **kwargs)
            # For some statuses, retry once, in case it's a transient problem.
            code = response.status_code
            if __debug__: log(addurl(f'got response with code {code}'))
            if code not in [400, 409, 502, 503, 504] or failures > _MAX_CONSECUTIVE_FAILS:
                return response
        except KeyboardInterrupt as ex:
            if __debug__: log(addurl(f'network {method} interrupted by {antiformat(ex)}'))
            raise
        except ImportError as ex:
            # This can happen if the installation environment has inconsistecies.
            log('import error: ' + str(ex))
            raise
        except (TypeError, httpx.UnsupportedProtocol) as ex:
            # Bad arguments to the call, like passing data to a 'get'.
            if __debug__: log(addurl(f'exception {antiformat(ex)}'))
            raise ArgumentError('Bad or invalid arguments in network call')
        except (httpx.CookieConflict, httpx.StreamError, httpx.TooManyRedirects,
                httpx.DecodingError, httpx.ProtocolError, httpx.ProxyError,
                httpx.ConnectError) as ex:
            # Probably indicates a deeper issue.  Don't do our lengthy retry
            # sequence, but try one more time, in case it's transient.
            if __debug__: log(addurl(f'exception {antiformat(ex)}'))
            if failures > 0:
                raise
            if __debug__: log(addurl('will retry one more time after brief pause'))
        except Exception as ex:         # noqa PIE786
            # Problem might be transient.  Don't quit right away.
            failures += 1
            if __debug__: log(addurl(f'exception (failure #{failures}): {antiformat(ex)}'))
            # Record the first error we get, not the subsequent ones, because
            # in the case of network outages, the subsequent ones will be
            # about being unable to reconnect and not the original problem.
            if not error:
                error = ex
        # If there's response text, it may contain diagnostic info.
        if __debug__ and response:
            text = response.text[:500] + (' ...' if len(response.text) > 500 else '')
            log('response text: ' + text)
        if failures == 0:
            # Pause briefly b/c it's rarely a good idea to retry immediately.
            if __debug__: log(addurl('pausing briefly before retrying'))
            wait(0.5)
            failures += 1
        elif failures < _MAX_CONSECUTIVE_FAILS:
            # Pause with exponential back-off, reset failure count & try again.
            pause = 5 * failures * failures
            if __debug__: log(addurl('pausing due to consecutive failures'))
            wait(pause)
            failures += 1
        else:
            if __debug__: log(addurl('exceeded max failures'))
            if error:
                raise error
            else:
                return response
    if interrupted():
        if __debug__: log(addurl('interrupted'))
        raise Interrupted(addurl('Network request has been interrupted'))
    else:
        # In theory, we should never reach this point.  If we do, then:
        raise InternalError(addurl('Unexpected case in timed_request'))


def net(method, url, client=None, handle_rate=True,
        polling=False, recursing=0, **kwargs):
    '''Invoke HTTP "method" on 'url' with optional keyword arguments provided.

    Returns a tuple of (response, exception), where the first element is
    the response from the get or post http call, and the second element is
    an exception object if an exception occurred.  If no exception occurred,
    the second element will be None.  This allows the caller to inspect the
    response even in cases where exceptions are raised.

    If keyword 'client' is not None, it's assumed to be a Python HTTPX Client
    object to use for the network call.  Settings such as timeouts should be
    done by the caller creating appropriately-configured Client objects.

    If keyword 'handle_rate' is True, this function will automatically pause
    and retry if it receives an HTTP code 429 ("too many requests") from the
    server.  If False, it will return the exception RateLimitExceeded instead.

    If keyword 'polling' is True, certain statuses like 404 are ignored and
    the response is returned; otherwise, they are considered errors.  The
    behavior when True is useful in situations where a URL does not exist until
    something is ready at the server, and the caller is repeatedly checking
    the URL.  It is up to the caller to implement the polling schedule and
    call this function (with polling = True) as needed.

    This method always passes the argument follow_redirects = True to the
    underlying Python HTTPX library network calls.
    '''
    import httpx

    if method.lower() not in _KNOWN_HTTP_METHODS:
        raise ValueError(f'Method must be one of {", ".join(_KNOWN_HTTP_METHODS)}.')

    def info(text, details=''):
        msg = f'{text} for {url}'
        if details:
            msg += (' (' + details + ')')
        return msg

    resp = None
    try:
        resp = timed_request(method, url, client, follow_redirects=True, **kwargs)
    except (httpx.NetworkError, httpx.ProtocolError) as ex:
        # timed_request() will have retried, so if we get here, time to bail.
        if __debug__: log(info(f'network exception: {antiformat(ex)}'))
        is_on_localhost = on_localhost(url)
        reason = antiformat(ex)
        if isinstance(ex, httpx.ConnectError) and is_on_localhost:
            if __debug__: log(info('returning ServiceFailure'))
            return (resp, ServiceFailure(info(f'Access failure ({reason})')))
        elif is_on_localhost or network_available():
            if __debug__: log(info('returning ServiceFailure'))
            return (resp, ServiceFailure(info(f'Server error ({reason})')))
        else:
            if __debug__: log(info('returning NetworkFailure'))
            return (resp, NetworkFailure(info('Network failure ({reason})')))
    except Exception as ex:             # noqa PIE786
        # Not a network or protocol error, and not a normal server response.
        if __debug__: log(info(f'returning exception: {antiformat(ex)}'))
        return (resp, ex)

    # Interpret the response.  Note that the requests library handles code 301
    # and 302 redirects automatically, so we don't need to do it here.
    error = None
    code = resp.status_code
    reason = resp.reason_phrase
    text = resp.text                    # Sometimes this holds more details.
    if code == 400:
        error = ServiceFailure(info('Server rejected the request', text))
    elif code in [401, 402, 403, 407, 451, 511]:
        error = AuthenticationFailure(info('Access is forbidden', text))
    elif code in [404, 410] and not polling:
        error = NoContent(info("No content found", text))
    elif code in [405, 406, 409, 411, 412, 413, 414, 417, 428, 431, 505, 510]:
        error = InternalError(info(f'Server returned code {code} ({reason})', text))
    elif code in [415, 416]:
        error = ServiceFailure(info(f'Server rejected the request ({reason})', text))
    elif code == 429:
        if handle_rate and recursing < _MAX_RECURSIVE_CALLS:
            pause = 5 * (recursing + 1)   # +1 b/c we start with recursing = 0.
            if __debug__: log(info('rate limit hit -- pausing', text))
            wait(pause)                   # 5 s, then 10 s, then 15 s, etc.
            if __debug__: log(info(f'doing recursive call #{recursing + 1}', text))
            return net(method, url, client, handle_rate, polling, recursing + 1, **kwargs)
        error = RateLimitExceeded(info('Server blocking requests due to rate limits', text))
    elif code in [500, 501, 502, 503, 504, 506, 507, 508]:
        error = ServiceFailure(info(f'Server error (code {code} -- {reason})', text))
    elif not (200 <= code < 400):
        error = NetworkFailure(info(f'Unable to resolve {url}', text))
    # The error msg will have had the URL added already; no need to do it here.
    if __debug__:                       # noqa SIM102
        if error:
            log('returning error ' + str(error))
    return (resp, error)


def network(method, url, client=None, handle_rate=True,
            polling=False, recursing=0, **kwargs):
    '''Invoke HTTP "method" on 'url' with optional keyword arguments provided.

    This is an alternative to net(). The difference is that this function only
    returns one value (the response object from HTTPX); if any error occurs,
    it raises an exception. (Compare this to net(...), which returns 2 values.)
    '''
    response, error = net(method, url, client, handle_rate,
                          polling, recursing, **kwargs)
    if error:
        raise error
    return response


def download_file(url, local_destination):
    '''Returns True if the content at 'url' could be downloaded to the file
    'local_destination', and False otherwise. It does not throw an exception.'''

    def addurl(text):
        return f'{text} for {url}'

    try:
        download(url, local_destination)
    except Exception as ex:             # noqa PIE786
        if __debug__: log(f'download exception: {antiformat(ex)}')
        return False
    else:
        return True


def download(url, local_destination, recursing=0):
    '''Download the 'url' to the file 'local_destination'.'''
    import httpx

    def addurl(text):
        return f'{text} for {url}'

    timeout = httpx.Timeout(15, connect=15, read=15, write=15)
    with httpx.stream('get', url, verify=False, timeout=timeout,
                      follow_redirects=True) as resp:
        code = resp.status_code
        if code == 202:
            # Code 202 = Accepted, "received but not yet acted upon."
            wait(2)                     # Sleep a short time and try again.
            raise_for_interrupts()
            recursing += 1
            if recursing <= _MAX_RECURSIVE_CALLS:
                if __debug__: log('calling download(url) recursively for code 202')
                download(url, local_destination, recursing)
            else:
                raise ServiceFailure(addurl('Exceeded max retries for code 202'))
        elif 200 <= code < 400:
            with open(local_destination, 'wb') as f:
                for chunk in resp.iter_bytes():
                    raise_for_interrupts()
                    f.write(chunk)
            resp.close()
            size = stat(local_destination).st_size
            if __debug__: log(f'wrote {size} bytes to file {local_destination}')
        elif code in [401, 402, 403, 407, 451, 511]:
            raise AuthenticationFailure(addurl('Access is forbidden'))
        elif code in [404, 410]:
            raise NoContent(addurl('No content found'))
        elif code in [405, 406, 409, 411, 412, 414, 417, 428, 431, 505, 510]:
            raise InternalError(addurl(f'Server returned code {code}'))
        elif code in [415, 416, 422]:
            raise ServiceFailure(addurl('Server rejected the request'))
        elif code == 429:
            raise RateLimitExceeded('Server blocking further requests due to rate limits')
        elif code == 503:
            raise ServiceFailure('Server is unavailable -- try again later')
        elif code in [500, 501, 502, 506, 507, 508]:
            raise ServiceFailure(addurl(f'Internal server error (HTTP code {code})'))
        else:
            raise NetworkFailure(f'Unable to resolve {url}')
